### PR TITLE
Optionally bind elements before appending to Arrays

### PR DESCRIPTION
**Pull request by ArtSabintsev**
This is a bit safer than the force-unwrap solution.

### CODE REVIEW

File: Sources/Utility/ArrayExtensions.swift [hunk 1]
```diff
@@ -20,10 +20,10 @@ extension Array {
...
         var u = [U]()
         for e in self {
-            if e is T {
-                t.append(e as! T)
-            } else {
-                u.append(e as! U)
+            if let e = e as? T {
+                t.append(e)
+            } else if let e = e as? U {
+                u.append(e)
```
Comment (aciidb0mb3r) [inline Sources/Utility/ArrayExtensions.swift]: how about 
```swift
            if case let e as T = e {
                t.append(e)
            } else if case let e as U = e {
                u.append(e)
            }
```

Does the same thing but avoids the `?` in `as`

Comment (ArtSabintsev) [inline Sources/Utility/ArrayExtensions.swift]: sure - that works!

**Comment by kostiakoval** [comment]
as for me `if case let` is to heavy

File: Sources/Utility/ArrayExtensions.swift (hunk 1 above)
Comment (kostiakoval) [inline Sources/Utility/ArrayExtensions.swift]: as for me if case let is to heavy

**Comment by ArtSabintsev** [comment]
actually, that does not work at all. Still asks to force-downcast. Agree with @kostiakoval.

File: Sources/Utility/ArrayExtensions.swift (hunk 1 above)
Comment (ArtSabintsev) [inline Sources/Utility/ArrayExtensions.swift]: actually, that does not work at all. Still asks to force-downcast. Agree with @kostiakoval.

Comment (aciidb0mb3r) [inline Sources/Utility/ArrayExtensions.swift]: Works for me
```swift
extension Array {
    public func partition<T, U>() -> ([T], [U]) {
        var t = [T]()
        var u = [U]()
        for e in self {
            if case let e as T = e {
                t.append(e)
            } else if case let e as U = e {
                u.append(e)
            }
        }
        return (t, u)
    }
}

let a = [1, 2, 3, "2", "3"]
let partition: ([Int], [String]) = a.partition()
```

agreed though that `if case let` looks heavier at first but I have grown to like it lol

Comment (kostiakoval) [inline Sources/Utility/ArrayExtensions.swift]: This works but I don's see how it's better than initial proposal. 
```swift 
if case let e as T = e {
    t.append(e)
} else if case let e as U = e {
    u.append(e)
}
```

Comment (aciidb0mb3r) [inline Sources/Utility/ArrayExtensions.swift]: wouldn't say better or worse just different syntactically I guess?

Comment (ArtSabintsev) [inline Sources/Utility/ArrayExtensions.swift]: @aciidb0mb3r's solution is syntactic sugar (https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Patterns.html#//apple_ref/swift/grammar/optional-pattern). Same solution, different way of implementing, which I would deem non-traditional.

Comment (kostiakoval) [inline Sources/Utility/ArrayExtensions.swift]: Both solutions works fine and there is no big difference. `if let e = e as? T ` solutions is safe, very clear and it's very often used in Swift community.
I feel there is no need for further discussion.

Comment (aciidb0mb3r) [inline Sources/Utility/ArrayExtensions.swift]: 👍

Comment (ArtSabintsev) [inline Sources/Utility/ArrayExtensions.swift]: Thanks! :+1:

**Pull request merged**